File: app/src/main/java/wangdaye/com/geometricweather/ui/widget/trend/TrendRecyclerView.java
package wangdaye.com.geometricweather.ui.widget.trend;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;

import androidx.annotation.ColorInt;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import android.util.AttributeSet;

import java.util.ArrayList;
import java.util.List;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.ui.widget.trend.item.AbsTrendItemView;
import wangdaye.com.geometricweather.utils.DisplayUtils;

/**
 * Trend recycler view.
 * */

public class TrendRecyclerView extends RecyclerView {

    private Paint mPaint;
    @ColorInt private int mLineColor;

    private int mDrawingBoundaryTop;
    private int mDrawingBoundaryBottom;

    private @Nullable List<KeyLine> mKeyLineList;
    private boolean mKeyLineVisibility = true;

    private @Nullable Float mHighestData;
    private @Nullable Float mLowestData;

    private int mTextSize;
    private int mTextMargin;
    private int mLineWidth;

    private static final int LINE_WIDTH_DIP = 1;
    private static final int TEXT_SIZE_DIP = 10;
    private static final int TEXT_MARGIN_DIP = 2;
    public static final int ITEM_MARGIN_BOTTOM_DIP = 16;

    public static class KeyLine {

        float value;
        String contentLeft;
        String contentRight;
        ContentPosition contentPosition;

        public enum ContentPosition {ABOVE_LINE, BELOW_LINE}

        public KeyLine(float value, String contentLeft, String contentRight, ContentPosition contentPosition) {
            this.value = value;
            this.contentLeft = contentLeft;
            this.contentRight = contentRight;
            this.contentPosition = contentPosition;
        }
    }

    public TrendRecyclerView(Context context) {
        super(context);
        initialize();
    }

    public TrendRecyclerView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        initialize();
    }

    public TrendRecyclerView(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initialize();
    }

    private void initialize() {
        setWillNotDraw(false);

        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setStrokeCap(Paint.Cap.ROUND);

        mTextSize = (int) DisplayUtils.dpToPx(getContext(), TEXT_SIZE_DIP);
        mTextMargin = (int) DisplayUtils.dpToPx(getContext(), TEXT_MARGIN_DIP);
        mLineWidth = (int) DisplayUtils.dpToPx(getContext(), LINE_WIDTH_DIP);

        mDrawingBoundaryTop = -1;
        mDrawingBoundaryBottom = -1;

        setLineColor(Color.GRAY);

        mKeyLineList = new ArrayList<>();
    }

    @Override
    public void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        drawKeyLines(canvas);
    }

    private void drawKeyLines(Canvas canvas) {
        if (!mKeyLineVisibility
                || mKeyLineList == null
                || mKeyLineList.size() == 0
                || mHighestData == null
                || mLowestData == null) {
            return;
        }

        if (getChildCount() > 0) {
            mDrawingBoundaryTop = ((AbsTrendItemView) getChildAt(0)).getChartTop();
            mDrawingBoundaryBottom = ((AbsTrendItemView) getChildAt(0)).getChartBottom();
        }
        if (mDrawingBoundaryTop < 0 || mDrawingBoundaryBottom < 0) {
            return;
        }

        float dataRange = mHighestData - mLowestData;
        float boundaryRange = mDrawingBoundaryBottom - mDrawingBoundaryTop;
        for (KeyLine line : mKeyLineList) {
            if (line.value > mHighestData || line.value < mLowestData) {
                continue;
            }

            int y = (int) (mDrawingBoundaryBottom - (line.value - mLowestData) / dataRange * boundaryRange);

            mPaint.setStyle(Paint.Style.STROKE);
            mPaint.setStrokeWidth(mLineWidth);
            mPaint.setColor(mLineColor);
            canvas.drawLine(0, y, getMeasuredWidth(), y, mPaint);

            mPaint.setStyle(Paint.Style.FILL);
            mPaint.setTextSize(mTextSize);
            mPaint.setColor(ContextCompat.getColor(getContext(), R.color.colorTextGrey2nd));
            switch (line.contentPosition) {
                case ABOVE_LINE:
                    mPaint.setTextAlign(Paint.Align.LEFT);
                    canvas.drawText(
                            line.contentLeft,
                            2 * mTextMargin,
                            y - mPaint.getFontMetrics().bottom - mTextMargin,
                            mPaint
                    );
                    mPaint.setTextAlign(Paint.Align.RIGHT);
                    canvas.drawText(
                            line.contentRight,
                            getMeasuredWidth() - 2 * mTextMargin,
                            y - mPaint.getFontMetrics().bottom - mTextMargin,
                            mPaint
                    );
                    break;

                case BELOW_LINE:
                    mPaint.setTextAlign(Paint.Align.LEFT);
                    canvas.drawText(
                            line.contentLeft,
                            2 * mTextMargin,
                            y - mPaint.getFontMetrics().top + mTextMargin,
                            mPaint
                    );
                    mPaint.setTextAlign(Paint.Align.RIGHT);
                    canvas.drawText(
                            line.contentRight,
                            getMeasuredWidth() - 2 * mTextMargin,
                            y - mPaint.getFontMetrics().top + mTextMargin,
                            mPaint
                    );
                    break;
            }
        }
    }

    // control.

    public void setData(List<KeyLine> keyLineList, float highestData, float lowestData) {
        mKeyLineList = keyLineList;
        mHighestData = highestData;
        mLowestData = lowestData;
        invalidate();
    }

    public void setKeyLineVisibility(boolean visibility) {
        mKeyLineVisibility = visibility;
        invalidate();
    }

    public void setLineColor(@ColorInt int lineColor) {
        mLineColor = lineColor;
        invalidate();
    }
}

File: app/src/main/java/org/breezyweather/ui/common/widgets/trend/TrendRecyclerView.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.common.widgets.trend

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import androidx.annotation.ColorInt
import androidx.recyclerview.widget.RecyclerView
import org.breezyweather.R
import org.breezyweather.common.extensions.dpToPx
import org.breezyweather.common.extensions.getTypefaceFromTextAppearance
import org.breezyweather.ui.common.widgets.trend.item.AbsTrendItemView

/**
 * Trend recycler view.
 */
class TrendRecyclerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0,
) : RecyclerView(context, attrs, defStyle) {
    private val mPaint = Paint().apply {
        isAntiAlias = true
        strokeCap = Paint.Cap.ROUND
    }

    @ColorInt
    private var mLineColor = 0
    private var mTextColor = 0
    private var mDrawingBoundaryTop: Int
    private var mDrawingBoundaryBottom: Int
    private var mKeyLineList: List<KeyLine>? = null
    private var mKeyLineVisibility = true
    private var mHighestData: Float? = null
    private var mLowestData: Float? = null
    private val mTextSize: Int
    private val mTextMargin: Int
    private val mLineWidth: Int

    class KeyLine(
        var value: Float,
        var contentLeft: String?,
        var contentRight: String?,
        var contentPosition: ContentPosition,
    ) {
        enum class ContentPosition {
            ABOVE_LINE,
            BELOW_LINE,
        }
    }

    init {
        setWillNotDraw(false)
        mPaint.typeface = getContext().getTypefaceFromTextAppearance(R.style.subtitle_text)
        mTextSize = getContext().dpToPx(TEXT_SIZE_DIP.toFloat()).toInt()
        mTextMargin = getContext().dpToPx(TEXT_MARGIN_DIP.toFloat()).toInt()
        mLineWidth = getContext().dpToPx(LINE_WIDTH_DIP.toFloat()).toInt()
        mDrawingBoundaryTop = -1
        mDrawingBoundaryBottom = -1
        setLineColor(Color.GRAY)
        setTextColor(Color.GRAY)
        mKeyLineList = mutableListOf()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawKeyLines(canvas)
    }

    private fun drawKeyLines(canvas: Canvas) {
        if (!mKeyLineVisibility ||
            mKeyLineList == null ||
            mKeyLineList!!.isEmpty() ||
            mHighestData == null ||
            mLowestData == null
        ) {
            return
        }
        if (childCount > 0) {
            mDrawingBoundaryTop = (getChildAt(0) as AbsTrendItemView).chartTop
            mDrawingBoundaryBottom = (getChildAt(0) as AbsTrendItemView).chartBottom
        }
        if (mDrawingBoundaryTop < 0 || mDrawingBoundaryBottom < 0) {
            return
        }
        val dataRange = mHighestData!! - mLowestData!!
        val boundaryRange = (mDrawingBoundaryBottom - mDrawingBoundaryTop).toFloat()
        for (line in mKeyLineList!!) {
            if (line.value > mHighestData!! || line.value < mLowestData!!) {
                continue
            }
            val y = (mDrawingBoundaryBottom - (line.value - mLowestData!!) / dataRange * boundaryRange).toInt()
            mPaint.apply {
                style = Paint.Style.STROKE
                strokeWidth = mLineWidth.toFloat()
                color = mLineColor
            }
            canvas.drawLine(0f, y.toFloat(), measuredWidth.toFloat(), y.toFloat(), mPaint)
            mPaint.apply {
                style = Paint.Style.FILL
                textSize = mTextSize.toFloat()
                color = mTextColor
            }
            when (line.contentPosition) {
                TrendRecyclerView.KeyLine.ContentPosition.ABOVE_LINE -> {
                    if (!line.contentLeft.isNullOrEmpty()) {
                        mPaint.textAlign = Paint.Align.LEFT
                        canvas.drawText(
                            line.contentLeft!!,
                            (2 * mTextMargin).toFloat(),
                            y - mPaint.fontMetrics.bottom - mTextMargin,
                            mPaint
                        )
                    }
                    if (!line.contentRight.isNullOrEmpty()) {
                        mPaint.textAlign = Paint.Align.RIGHT
                        canvas.drawText(
                            line.contentRight!!,
                            (measuredWidth - 2 * mTextMargin).toFloat(),
                            y - mPaint.fontMetrics.bottom - mTextMargin,
                            mPaint
                        )
                    }
                }

                TrendRecyclerView.KeyLine.ContentPosition.BELOW_LINE -> {
                    if (!line.contentLeft.isNullOrEmpty()) {
                        mPaint.textAlign = Paint.Align.LEFT
                        canvas.drawText(
                            line.contentLeft!!,
                            (2 * mTextMargin).toFloat(),
                            y - mPaint.fontMetrics.top + mTextMargin,
                            mPaint
                        )
                    }
                    if (!line.contentRight.isNullOrEmpty()) {
                        mPaint.textAlign = Paint.Align.RIGHT
                        canvas.drawText(
                            line.contentRight!!,
                            (measuredWidth - 2 * mTextMargin).toFloat(),
                            y - mPaint.fontMetrics.top + mTextMargin,
                            mPaint
                        )
                    }
                }
            }
        }
    }

    // control.
    fun setData(keyLineList: List<KeyLine>?, highestData: Float, lowestData: Float) {
        mKeyLineList = keyLineList
        mHighestData = highestData
        mLowestData = lowestData
        invalidate()
    }

    fun setKeyLineVisibility(visibility: Boolean) {
        mKeyLineVisibility = visibility
        invalidate()
    }

    fun setLineColor(@ColorInt lineColor: Int) {
        mLineColor = lineColor
        invalidate()
    }

    fun setTextColor(@ColorInt textColor: Int) {
        mTextColor = textColor
        invalidate()
    }

    companion object {
        private const val LINE_WIDTH_DIP = 1
        private const val TEXT_SIZE_DIP = 12
        private const val TEXT_MARGIN_DIP = 2
        const val ITEM_MARGIN_BOTTOM_DIP = 16
    }
}
